tests/test.ZoneeventWatcherOverflow.js

```diff
@@ -0,0 +1,224 @@
+var vmEvents = {};
+
+
+// "CONFIG"
+var CHARS_PER_LINE = 200;  // at least this many chars in a zoneevent line
+var STRAGGLER_WAIT = 5000; // wait this many ms for events after delete
+var OVERHEAD_EVENTS = 15;  // events we expect for create/delete
+var RESTART_EVENTS = 9;    // events we expect for each restart
+var NUM_RESTARTS_PER_VM = 25;
+var NUM_VMS = 10;
+
+
+/*
+ * This test exists to confirm that AGENT-987 is fixed.
+ *
+ * Basically it proves that we can successfully handle at least (NUM_VMS *
+ * NUM_RESTARTS_PER_VM * 9) + (NUM_VMS * 15) events.
+ *
+ * To test this, we create NUM_VMS VMs and reboot them in parallel
+ * NUM_RESTARTS_PER_VM times.
+ *
+ * This should result in that number of start + stop events for each of them
+ * along with 7 other intermediate events per restart and 15 events per VM for
+ * the create/destroy.
+ *
+ */
+
+test('find SmartOS image', function _test(t) {
+    common.testFindSmartosImage(t, function _findSmartosCb(err, latest) {
+        t.ifError(err, 'find SmartOS Image');
+        if (err) {
+            throw new Error('Cannot continue without SmartOS Image');
+        }
+        smartosImageUUID = latest;
+        t.end();
+    });
+});
+
+test('starting ZoneeventWatcher', function _test(t) {
+    function _onVmUpdate(vmUuid, updateType, updateObj) {
+        if (!vmEvents.hasOwnProperty(vmUuid)) {
+            vmEvents[vmUuid] = {start: 0, stop: 0, other: 0};
+        }
+
+        if (updateObj.zone_state === 'running') {
+            vmEvents[vmUuid].start++;
+        } else if (updateObj.zone_state === 'stopped') {
+            vmEvents[vmUuid].stop++;
+        } else {
+            vmEvents[vmUuid].other++;
+        }
+    }
+
+    watcher = new ZoneeventWatcher({
+        debugEvents: true,
+        highWaterMark: 32, // default is 16k! takes too long to get there.
+        log: mocks.Logger,
+        updateVm: _onVmUpdate
+    });
+
+    watcher.start();
+    t.ok(watcher, 'created ZoneeventWatcher [' + watcher.getPid() + ']');
+
+    t.end();
+});
+
+test('create VMs', function _test(t) {
+    var numVms = NUM_VMS;
+    var vmsToCreate = [];
+
+    while (numVms > 0) {
+        vmsToCreate.push(numVms--);
+    }
+
+    function _createVm(idx, cb) {
+        var payload = {
+            alias: 'vm-agent_testvm-overflow-' + idx,
+            autoboot: true,
+            brand: 'joyent-minimal',
+            image_uuid: smartosImageUUID,
+            quota: 10
+        };
+
+        payload.log = mocks.Logger;
+
+        vmadm.create(payload, function _vmadmCreateCb(err, info) {
+            t.ifError(err, 'create VM'
+                + ((!err && info) ? ': ' + info.uuid : ''));
+            if (!err && info) {
+                createdVms.push(info.uuid);
+            }
+            cb(err);
+        });
+    }
+
+    vasync.forEachParallel({
+        inputs: vmsToCreate,
+        func: _createVm
+    }, function _afterForEachParallel(err) {
+        t.ifError(err, 'done creating');
```

Comment:
Yes, that sounds much better. Although, wouldn't `VM` need to be at the plural form? If I understand correctly, we're creating several VMs.